server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -900,15 +912,58 @@ private void asyncCleanupUnlinkedShardLevelBlobs(
-            try {
-                deleteFromContainer(blobContainer(), filesToDelete);
-                l.onResponse(null);
-            } catch (Exception e) {
-                logger.warn(() -> new ParameterizedMessage("{} Failed to delete some blobs during snapshot delete", snapshotIds), e);
-                throw e;
+
+        final BlockingQueue<List<String>> staleFilesToDeleteInBatch = new LinkedBlockingQueue<>();
+        final List<String> partition = new ArrayList<>();
+
+        try {
+            for (String key : filesToDelete) {
+                partition.add(key);
+                if (maxStaleBlobDeleteBatch == partition.size()) {
+                    staleFilesToDeleteInBatch.add(new ArrayList<>(partition));
+                    partition.clear();
+                }
             }
-        }));
+            if (partition.isEmpty() == false) {
+                staleFilesToDeleteInBatch.add(new ArrayList<>(partition));
+            }
+
+            final GroupedActionListener<Void> groupedListener = new GroupedActionListener<>(
+                ActionListener.wrap(r -> { listener.onResponse(null); }, listener::onFailure),
+                staleFilesToDeleteInBatch.size()
+            );
+
+            // Start as many workers as fit into the snapshot pool at once at the most
+            final int workers = Math.min(threadPool.info(ThreadPool.Names.SNAPSHOT).getMax(), filesToDelete.size());
+            for (int i = 0; i < workers; ++i) {
+                executeStaleShardDelete(staleFilesToDeleteInBatch, groupedListener);
+            }
+
+        } catch (Exception e) {
+            // TODO: We shouldn't be blanket catching and suppressing all exceptions here and instead handle them safely upstream.
+            // Currently this catch exists as a stop gap solution to tackle unexpected runtime exceptions from implementations
+            // bubbling up and breaking the snapshot functionality.
+            assert false : e;
+            logger.warn(new ParameterizedMessage("[{}] Exception during cleanup of stale blobs", snapshotIds), e);
+        }
+    }
+
+    private void executeStaleShardDelete(BlockingQueue<List<String>> staleFilesToDeleteInBatch, GroupedActionListener<Void> listener)
+        throws InterruptedException {
+        List<String> filesToDelete = staleFilesToDeleteInBatch.poll(0L, TimeUnit.MILLISECONDS);
+        if (filesToDelete != null) {
+            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(listener, l -> {
+                try {
+                    deleteFromContainer(blobContainer(), filesToDelete);
+                    l.onResponse(null);
+                } catch (Exception e) {
+                    logger.warn(() -> new ParameterizedMessage("{} Failed to delete blobs during snapshot delete", metadata.name()), e);
+                    throw e;
+                }
+            }));
+
+            executeStaleShardDelete(staleFilesToDeleteInBatch, listener);
```

Comment:
This is the part missing for the work-stealing method used throughout this file. Why did you choose to have the recursion outside the initial thread? this is a (minor) problem since it doesn't correspond with the `for` loop initializing the threads.
```suggestion
            
            executeStaleShardDelete(staleFilesToDeleteInBatch, listener);
            }));
```

Also, this means you would need to change the way you handle errors. The exception would terminate the thread, so potentially you could reach a situation where the `GroupedActionListener` is waiting forever and the process gets stuck, for example:
- X threads start running 
- Eventually, they all get an exception and stop the recursion *before* the queue is empty!
- the `GroupedActionListener` is stuck waiting for the tasks to be done... 

I suggest not throwing the exception, but invoking `l.onFailue()` if there is an exception. this way the other shards may still be handled.

Reply:
`ActionRunnable` which is of type `AbstractRunnable` invokes the `l.onFailure(e)` when runnable throws any exception. Ref [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/util/concurrent/AbstractRunnable.java#L52)

Reply:
there is no reason not to simply invoke onFailure with the exception in it -> the same outcome, except that the exception would stop current recursion and process could hang.